pkg/volume/util/atomic_writer.go

```diff
@@ -113,9 +116,9 @@ const (
 //      The data directory itself is a link to a timestamped directory with
 //      the real data:
 //        <target-dir>/..data          -> ..2016_02_01_15_04_05.12345678/
-//  8.  A symlink to the new timestamped directory ..data_tmp is created that will
-//      become the new data directory
-//  9.  The new data directory symlink is renamed to the data directory; rename is atomic
+//      NOTE(claudiub): We need to create these symlinks AFTER we've finished creating and
+//      linking everything else. On Windows, if a target does not exist, the created symlink
+//      will not work properly if the target ends up being a directory.
```

Comment:
Could you provide some details what errors show up with the previous approach?

Reply:
I've included in the issue https://github.com/kubernetes/kubernetes/issues/97802 a pod that can replicate this issue. The same issue has been observed in 2 Conformance tests (indeed, they are marked as ``[LinuxOnly]``, but that was because we didn't support containerd at the time. The plan is to drop that ``[LinuxOnly]`` tag for some of the tests when dockershim will be dropped). The Conformance tests are:

```
[sig-storage] Subpath Atomic writer volumes should support subpaths with downward pod [LinuxOnly] [Conformance]
[sig-storage] Subpath Atomic writer volumes should support subpaths with projected pod [LinuxOnly] [Conformance]
```

The error encountered while spawning the pod:

```
Events:
  Type     Reason     Age        From               Message
  ----     ------     ----       ----               -------
  Normal   Scheduled  <invalid>  default-scheduler  Successfully assigned default/pod-subpath-test-downwardapi to 9932k8s000
  Normal   Pulling    <invalid>  kubelet            Pulling image "claudiubelu/agnhost:2.21"
  Normal   Pulled     <invalid>  kubelet            Successfully pulled image "claudiubelu/agnhost:2.21" in 28.9997536s
  Warning  Failed     <invalid>  kubelet            Error: failed to create containerd task: hcsshim::CreateComputeSystem test-container-subpath-downwardapi: The filename, directory name, or volume label syntax is incorrect.
(extra info: {"Owner":"containerd-shim-runhcs-v1.exe","SchemaVersion":{"Major":2,"Minor":1},"Container":{"GuestOs":{"HostName":"pod-subpath-test-downwardapi"},"Storage":{"Layers":[{"Id":"77cfc4d7-0118-5908-8826-ad4d330c755a","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8191"},{"Id":"44fb9d95-0c15-507d-bc04-3475e78fbfd4","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8190"},{"Id":"f4b2c539-6bf5-51d0-aa26-372861d511d0","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8189"},{"Id":"ea209b30-8df1-5907-be7c-813f76fc0a6a","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8188"},{"Id":"09565c90-2bdc-521b-94cb-5683931775b2","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8187"},{"Id":"7103248a-bd81-59c8-a5c3-3528a6fc36b8","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8186"},{"Id":"688853a5-9780-5721-8c19-d7a61c65deb4","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8185"},{"Id":"19795dd2-362e-50a6-be0c-c81525fde5ca","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8184"},{"Id":"99a7b2ce-a5b6-56d2-bd50-38929d8cc8c4","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8183"},{"Id":"13892c24-954b-5bce-b09d-b0ae0234ed49","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8182"},{"Id":"d89e572a-21fb-5f0b-88bc-10f221f85b19","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8181"},{"Id":"28524ec2-f0cc-5090-9ad4-8a1d2cb4f796","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8180"},{"Id":"c6512116-388b-5d4d-864b-50370f2f783d","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8179"},{"Id":"3776a196-9239-578b-b856-3e6a5e98829b","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8178"},{"Id":"0d1a434f-150e-5e58-bb6c-7fef2d0c22d8","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8177"},{"Id":"6fdc9067-9e47-559a-9879-cb1b269b03b1","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8176"},{"Id":"a18f5f64-5e5f-508f-ba98-468c6f2ae3bf","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8175"},{"Id":"e72d95de-818e-50c4-b50f-03b5cc9a2765","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8174"},{"Id":"d8da0836-8951-589f-b74e-6c6989fe23fd","Path":"C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\8173"}],"Path":"\\\\?\\Volume{c79e425a-19c7-4ba4-9f56-ffb7ee551b91}\\"},"MappedDirectories":[{"HostPath":"c:\\var\\lib\\kubelet\\pods\\14772bf0-a2d1-46a2-aa2d-d53bb07a658f\\volumes\\kubernetes.io~downward-api\\test-volume\\downward\\podname","ContainerPath":"c:\\test-volume","ReadOnly":true},{"HostPath":"c:\\var\\lib\\kubelet\\pods\\14772bf0-a2d1-46a2-aa2d-d53bb07a658f\\containers\\test-container-subpath-downwardapi\\e133dad5","ContainerPath":"c:\\dev\\termination-log"},{"HostPath":"c:\\var\\lib\\kubelet\\pods\\14772bf0-a2d1-46a2-aa2d-d53bb07a658f\\volumes\\kubernetes.io~projected\\kube-api-access-58449","ContainerPath":"c:\\var\\run\\secrets\\kubernetes.io\\serviceaccount","ReadOnly":true},{"HostPath":"c:\\var\\lib\\kubelet\\pods\\14772bf0-a2d1-46a2-aa2d-d53bb07a658f\\etc-hosts","ContainerPath":"C:\\Windows\\System32\\drivers\\etc\\hosts"}],"Networking":{"Namespace":"58fe8f97-666f-419a-ae11-7d7b00d2a5d5"}},"ShouldTerminateOnLastHandleClosed":true}): unknown
```

Using this PR will result in the scenario working, and the tests passing.

As mentioned in this PR description, on Windows, if you create a symlink to a target that does **not** exist yet, Windows will create a regular symlink. However, if afterwards the created target is actually a directory, the symlink will not work because it should have been a symlinkd, not a symlink.

Seems like this is a known issue, too: https://github.com/golang/go/issues/39183